File: tamiltts/phonemize.py
"""Tamil phonemization — the single source of truth shared by training and inference.

The trained VITS model maps a sequence of *phoneme ids* to audio. To run the exported ONNX
model outside Coqui-TTS (in our Python and Rust SDKs) we must reproduce the exact same
text -> phoneme-id pipeline used at training time. That contract is:

    1. espeak-ng converts Tamil graphemes -> IPA phonemes  (`espeak-ng -q --ipa=3 -v <lang>`)
    2. each IPA character is mapped to an id via the vocabulary in `tokenizer.json`
    3. if `add_blank` is set, a blank id is interleaved between phonemes and at both ends
    4. `bos`/`eos` ids are prepended/appended if present

`tokenizer.json` (written by `tamiltts.export_onnx`) is the authoritative description of the
vocabulary and these flags, so the Python and Rust SDKs stay byte-for-byte consistent.
"""

from __future__ import annotations

import json
import shutil
import subprocess
from dataclasses import dataclass, field
from pathlib import Path

DEFAULT_LANGUAGE = "ta"


def espeak_phonemes(text: str, language: str = DEFAULT_LANGUAGE) -> str:
    """Return the IPA phoneme string for `text` using the espeak-ng CLI.

    Uses ``--ipa=3`` (phonemes separated, no stress/tie ties) for stable, parseable output.
    """
    exe = shutil.which("espeak-ng") or shutil.which("espeak")
    if exe is None:
        raise RuntimeError(
            "espeak-ng not found on PATH. Install it: `brew install espeak-ng`."
        )
    out = subprocess.run(
        [exe, "-q", "--ipa=3", "-v", language, text],
        capture_output=True,
        text=True,
        check=True,
    )
    # espeak emits one line per sentence; join and normalize whitespace.
    return " ".join(out.stdout.split())


@dataclass
class Tokenizer:
    """Maps Tamil text -> phoneme ids, per an exported `tokenizer.json` contract."""

    id_map: dict[str, int]
    language: str = DEFAULT_LANGUAGE
    add_blank: bool = True
    blank_id: int = 0
    pad_id: int = 0
    bos_id: int | None = None
    eos_id: int | None = None
    # characters that espeak may emit but the model never saw; dropped silently.
    ignore: set[str] = field(default_factory=lambda: {" "})

    @classmethod
    def from_file(cls, path: str | Path) -> "Tokenizer":
        data = json.loads(Path(path).read_text(encoding="utf-8"))
        return cls(
            id_map={str(k): int(v) for k, v in data["id_map"].items()},
            language=data.get("language", DEFAULT_LANGUAGE),
            add_blank=bool(data.get("add_blank", True)),
            blank_id=int(data.get("blank_id", 0)),
            pad_id=int(data.get("pad_id", 0)),
            bos_id=data.get("bos_id"),
            eos_id=data.get("eos_id"),
            ignore=set(data.get("ignore", [" "])),
        )

    def to_dict(self) -> dict:
        return {
            "language": self.language,
            "add_blank": self.add_blank,
            "blank_id": self.blank_id,
            "pad_id": self.pad_id,
            "bos_id": self.bos_id,
            "eos_id": self.eos_id,
            "ignore": sorted(self.ignore),
            "id_map": self.id_map,
        }

    def encode(self, text: str) -> list[int]:
        phonemes = espeak_phonemes(text, self.language)
        ids: list[int] = []
        for ch in phonemes:
            if ch in self.ignore:
                continue
            tok = self.id_map.get(ch)
            if tok is not None:
                ids.append(tok)

        if self.add_blank:
            interleaved = [self.blank_id]
            for tok in ids:
                interleaved.append(tok)
                interleaved.append(self.blank_id)
            ids = interleaved

        if self.bos_id is not None:
            ids = [self.bos_id, *ids]
        if self.eos_id is not None:
            ids = [*ids, self.eos_id]
        return ids

File: rust/src/phonemize.rs
//! Tamil phonemization — the Rust half of the cross-language contract.
//!
//! Mirrors `tamiltts/phonemize.py`: shell out to the same `espeak-ng` binary, then map IPA
//! characters to ids using the same `tokenizer.json`. Keeping both SDKs on the identical
//! espeak CLI invocation is what makes the exported ONNX model portable across languages.

use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::process::Command;

use anyhow::{anyhow, Context, Result};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
pub struct Tokenizer {
    #[serde(default = "default_language")]
    pub language: String,
    #[serde(default = "default_true")]
    pub add_blank: bool,
    #[serde(default)]
    pub blank_id: i64,
    #[serde(default)]
    pub pad_id: i64,
    #[serde(default)]
    pub bos_id: Option<i64>,
    #[serde(default)]
    pub eos_id: Option<i64>,
    #[serde(default = "default_sample_rate")]
    pub sample_rate: u32,
    #[serde(default = "default_ignore")]
    pub ignore: Vec<String>,
    pub id_map: HashMap<String, i64>,
}

fn default_language() -> String {
    "ta".to_string()
}
fn default_true() -> bool {
    true
}
fn default_sample_rate() -> u32 {
    22050
}
fn default_ignore() -> Vec<String> {
    vec![" ".to_string()]
}

impl Tokenizer {
    pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self> {
        let data = std::fs::read_to_string(path.as_ref())
            .with_context(|| format!("reading tokenizer {}", path.as_ref().display()))?;
        let tok: Tokenizer = serde_json::from_str(&data).context("parsing tokenizer.json")?;
        Ok(tok)
    }

    /// Run espeak-ng to convert Tamil text to an IPA phoneme string (`--ipa=3`).
    fn espeak(&self, text: &str) -> Result<String> {
        let out = Command::new("espeak-ng")
            .args(["-q", "--ipa=3", "-v", &self.language, text])
            .output()
            .context("running espeak-ng (is it installed? `brew install espeak-ng`)")?;
        if !out.status.success() {
            return Err(anyhow!(
                "espeak-ng failed: {}",
                String::from_utf8_lossy(&out.stderr)
            ));
        }
        let raw = String::from_utf8_lossy(&out.stdout);
        Ok(raw.split_whitespace().collect::<Vec<_>>().join(" "))
    }

    /// Encode Tamil text to the phoneme-id sequence the ONNX model expects.
    pub fn encode(&self, text: &str) -> Result<Vec<i64>> {
        let phonemes = self.espeak(text)?;
        let ignore: HashSet<&str> = self.ignore.iter().map(|s| s.as_str()).collect();

        let mut ids: Vec<i64> = Vec::new();
        for ch in phonemes.chars() {
            let key = ch.to_string();
            if ignore.contains(key.as_str()) {
                continue;
            }
            if let Some(&id) = self.id_map.get(&key) {
                ids.push(id);
            }
        }

        if self.add_blank {
            let mut interleaved = Vec::with_capacity(ids.len() * 2 + 1);
            interleaved.push(self.blank_id);
            for id in ids {
                interleaved.push(id);
                interleaved.push(self.blank_id);
            }
            ids = interleaved;
        }
        if let Some(bos) = self.bos_id {
            ids.insert(0, bos);
        }
        if let Some(eos) = self.eos_id {
            ids.push(eos);
        }

        if ids.is_empty() {
            return Err(anyhow!("text produced no phonemes; check input / espeak language"));
        }
        Ok(ids)
    }
}
